src/permissionSelectors.js

```diff
@@ -40,6 +40,63 @@ export function roleIsAtLeast(thisRole: RoleT, thresholdRole: RoleT): boolean {
   return thisRole <= thresholdRole; // Roles with more privilege have lower numbers.
 }
 
+/**
+ * Whether the self-user is authorized to create or edit a stream to be
+ *   public.
+ *
+ * Note: This isn't about web-public streams. For those, see
+ * getCanCreateWebPublicStreams.
+ */
+export function getCanCreatePublicStreams(state: PerAccountState): boolean {
+  const { createPublicStreamPolicy } = getRealm(state);
+  const role = getOwnUserRole(state);
+
+  switch (createPublicStreamPolicy) {
+    case 4: // ModeratorOrAbove
+      return roleIsAtLeast(role, Moderator);
+    case 3: // FullMemberOrAbove
+      // TODO: Implement properly.
+      return roleIsAtLeast(role, Member);
+    case 2: // AdminOrAbove
+      return roleIsAtLeast(role, Admin);
+    case 1: // AdminOrAbove
+      return roleIsAtLeast(role, Member);
+    default: {
+      ensureUnreachable(createPublicStreamPolicy);
+
+      // (Unreachable as long as the cases are exhaustive.)
+      return false;
+    }
+  }
+}
+
+/**
+ * Whether the self-user is authorized to create or edit a stream to be
+ *   private.
+ */
+export function getCanCreatePrivateStreams(state: PerAccountState): boolean {
+  const { createPrivateStreamPolicy } = getRealm(state);
+  const role = getOwnUserRole(state);
+
+  switch (createPrivateStreamPolicy) {
+    case 4: // ModeratorOrAbove
+      return roleIsAtLeast(role, Moderator);
+    case 3: // FullMemberOrAbove
+      // TODO: Implement properly.
+      return roleIsAtLeast(role, Member);
+    case 2: // AdminOrAbove
+      return roleIsAtLeast(role, Admin);
+    case 1: // AdminOrAbove
+      return roleIsAtLeast(role, Member);
+    default: {
```

Comment:
I think the duplication of these `switch` statements is OK, but (together with the similar one in `getCanCreateWebPublicStreams`) it's also getting to the point where it'd be helpful to unify them.

In particular, CreatePublicOrPrivateStreamPolicy and CreateWebPublicStreamPolicy agree on the meaning of the values they have in common. That means a single switch can take a policy value that's in the union of those, and get the right answer regardless of which is intended.

In the web implementation, this logic is [`user_has_permission`](https://github.com/zulip/zulip/blob/134977b5903e4b1ceee14ea81c1988f6d677c073/static/js/settings_data.ts#L115),
and the policy values are in… three different places [in `settings_config.ts`](https://github.com/zulip/zulip/blob/134977b5903e4b1ceee14ea81c1988f6d677c073/static/js/settings_config.ts): `common_policy_values`, `common_message_policy_values`, and `create_web_public_stream_policy_values`.

Reply:
To avoid ambiguity:
* This duplication of switch statements is fine.
* Unifying the three of them would be a nice improvement, but isn't necessary. (And at this point probably isn't worth any significant effort -- better to call this feature area done for now, ship it, and move on to other things.)
* If we had like six of these instead of three, the duplication would become increasingly annoying, and unifying them increasingly useful.

Reply:
And then re the `permissionSelectors` tests: I think that duplication, reflecting the duplication that's in the app code, is fine. In a future where we dedupe this app logic, we can dedupe the test code naturally in a corresponding way.

Reply:
Cool, makes sense. I'll link to this discussion in code comments.